xdsl/transforms/dead_code_elimination.py

```diff
@@ -73,8 +79,80 @@ def dce(op: ModuleOp):
     walker.rewrite_module(op)
 
 
+@dataclass
+class LiveSet:
+    changed: bool = field(default=True)  # Force first iteration to happen
+    _live_ops: set[Operation] = field(default_factory=set)
+
+    def is_live(self, op: Operation) -> bool:
+        return op in self._live_ops
+
+    def set_live(self, op: Operation):
+        if not self.is_live(op):
+            self.changed = True
+            self._live_ops.add(op)
+
+    def propagate_op_liveness(self, op: Operation):
+        for region in op.regions:
+            self.propagate_region_liveness(region)
+
+        if self.is_live(op):
+            return
+
+        if not would_be_trivially_dead(op):
+            self.set_live(op)
+            return
+
+        if any(
+            self.is_live(use.operation) for result in op.results for use in result.uses
+        ):
+            self.set_live(op)
+
+    def propagate_region_liveness(self, region: Region):
+        first = region.first_block
+        if first is None:
+            return
+        for block in PostOrderIterator(first):
+            # Process operations in reverse order to speed convergence
+            for operation in reversed(block.ops):
+                self.propagate_op_liveness(operation)
+
+    def delete_dead(self, region: Region):
+        first = region.first_block
+        if first is None:
+            return
+
+        for block in reversed(region.blocks):
+            if not any(self.is_live(op) for op in block.ops) and block != first:
+                # If block is not the entry block and has no live ops then delete it
+                self.changed = True
+                region.erase_block(block, safe_erase=False)
+                continue
+
+            for operation in reversed(block.ops):
+                if not self.is_live(operation):
+                    self.changed = True
+                    block.erase_op(operation, safe_erase=False)
+                else:
+                    for r in operation.regions:
+                        self.delete_dead(r)
+
+
+def region_dce(region: Region) -> bool:
+    live_set = LiveSet()
+
+    while live_set.changed:
+        live_set.changed = False
+        live_set.propagate_region_liveness(region)
+
+    live_set.changed = False
+    live_set.delete_dead(region)
+    return live_set.changed
+
+
 class DeadCodeElimination(ModulePass):
     name = "dce"
 
     def apply(self, ctx: MLContext, op: ModuleOp) -> None:
-        dce(op)
+        for region in op.regions:
+            region_dce(region)
```

Comment:
This only does regions of the root operation, and not recursively, right?

Reply:
It would be good to add a test with a module op, that has a test op with a region with dead code

Reply:
This should act recursively, I'll add a test